src/graph/GraphRunCtx.hpp

```diff
@@ -33,27 +34,64 @@ struct GraphRunCtx
 	/**
 	 * Executes graph bound with this GraphRunCtx.
 	 */
-	void run();
+	void executeAsync();
 
 	/**
-	 * Resets graphRunCtx for all nodes currently bound with this GraphRunCtx, consequently destroying this GraphRunCtx.
-	 */
+	* Resets graphRunCtx for all nodes currently bound with this GraphRunCtx, consequently destroying this GraphRunCtx.
+	*/
 	void detachAndDestroy();
 
+	/**
+	 * Waits until this GraphRunCtx
+	 * - finishes execution
+	 * - joins its thread
+	 * - synchronizes graph stream (all pending GPU operations)
+	 */
+	void synchronize();
+
+	/**
+	 * Ensures that no GraphRunCtx is running.
+	 */
+	static void synchronizeAll();
+
+	/**
+	 * Waits until given node finishes its CPU execution.
+	 * The node may still have pending GPU operations.
+	 */
+	void synchronizeNodeCPU(Node::Ptr nodeToSynchronize);
+
+
 	CudaStream::Ptr getStream() const { return stream; }
 	const std::set<std::shared_ptr<Node>>& getNodes() const { return nodes; }
 
 	virtual ~GraphRunCtx();
 private:
-	GraphRunCtx() : stream(CudaStream::create()) {}
+	GraphRunCtx() : stream(CudaStream::create(cudaStreamNonBlocking)) {}
 
 	static std::vector<std::shared_ptr<Node>> findExecutionOrder(std::set<std::shared_ptr<Node>> nodes);
 
+	void executeThreadMain();
+
 private:
 	CudaStream::Ptr stream;
-	std::set<std::shared_ptr<Node>> nodes;
-	std::vector<std::shared_ptr<Node>> executionOrder;
+	std::optional<std::thread> maybeThread;
+	std::set<Node::Ptr> nodes;
+	std::vector<Node::Ptr> executionOrder;
+
+	// Used to synchronize all existing instances (e.g. to safely access Scene).
+	static std::list<std::weak_ptr<GraphRunCtx>> instances;
```

Comment:
Both solutions can be justified. I decided to move it to `Node` because it is frequently used there, and having a shared_ptr instead of weak_ptr removes all `weak_ptr::lock()` overhead.